### PR TITLE
Migrating access token from deprecated storage to AccountStores

### LINKED ISSUE #4394: FluxC integration checklist
Stores integration:
- [x] Integrate AccountStore - @maxme
- [x] Integrate SiteStore - @maxme
- [x] Integrate PostStore - @aforcier
- [ ] Integrate MediaStore - @tonyr59h / @oguzkocer 
- [x] Integrate CommentStore - @maxme
- [x] Integrate TaxonomyStore - @aforcier
- [ ] Rewrite `MediaUploadService` and `PostUploadService`

Misc:
- [x] Access token migration to the AccountStore from any wpandroid version - @tonyr59h
- [x] Integrate discovery to the sign in screen (and to the add new site screen) - @maxme
- [x] Integrate HTTP Auth and SSL certificate management - @maxme
- [x] Integrate 2FA - @tonyr59h
- [x] Migrate .org sites from older app version #5027
- [ ] Migrate local drafts from older app version (@aforcier)
- [ ] Make sure some key analytics are still there and well tracked (sign in / sign in errors for instance).
- [x] Kill SelfSignedSSLCertsManager (ref https://github.com/wordpress-mobile/WordPress-Android/pull/4395)
- [x] "Detect" self hosted sites with Jetpack not added via .com login and ask the user to login if they want to access specific screens (Stats, themes, site settings, ?)

### PR DESCRIPTION
**Pull request by tonyr59h**
Partially Fixes #4394

The migration process first looks at the `AccountsTable` (the most recent) then in `SharedPreferences`.

If a token is found the AccountStore token is updated and the local `OAuthAuthenticator` token is set. If no token is found the local `OAuthAuthenticator` token will be set to the current value held in the AccountStore.

Needs review: @maxme

### CODE REVIEW

**Comment by maxme** [comment]
We should actually try to get the access token from the `AccountTable` where it lives actually in `develop`: https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/datasets/AccountTable.java#L34

And eventually from the old deprecated `accounts` table and `preferences`, if it wasn't found.

**Comment by maxme** [comment]
- [ ] Migration test: checkout an old version (5.6.1 for instance), sign in a wpcom account - then checkout this branch, build and install, make sure you're still logged in.

**Comment by maxme** [comment]
Writing the previous test made me think about something: we will have to run a few actions (FETCH_ACCOUNT_INFOS, FETCH_ACCOUNT_SETTINGS, and FETCH_SITES) only if a migration was applied.

**Comment by tonyr59h** [comment]
> We should actually try to get the access token from the AccountTable where it lives actually in develop

_Sigh_, missed that one. Checking that table in dc33e8a30ce8e8290c48dd8d95e52440ed8ecbec.

**Comment by tonyr59h** [comment]
> Writing the previous test made me think about something

Did you write this test already?

**Comment by maxme** [comment]
> Did you write this test already?

Nope, I meant "Writing the previous test scenario" ;)

File: WordPress/src/main/java/org/wordpress/android/WordPress.java [hunk 1]
```diff
@@ -972,11 +973,11 @@ private String migrateAccessTokenToAccountStore() {
         return token;
     }
 
-    private String getDeprecatedAccountTableAccessToken() {
+    private String getAccessTokenFromTable(String tableName) {
         String token = null;
         try {
             SQLiteDatabase db = SQLiteDatabase.openDatabase(DEPRECATED_DATABASE_NAME, null, SQLiteDatabase.OPEN_READONLY);
```
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/WordPress.java]: This line doesn't work, we should use:
```java
SQLiteDatabase db = getApplicationContext().openOrCreateDatabase(DEPRECATED_DATABASE_NAME, 0, null);
```

**Comment by maxme** [comment]
We should move the migration code to a separate utils class.

**Comment by maxme** [comment]
We should only do the migration if there isn't a token in the AccountStore.

**Comment by tonyr59h** [comment]
> We should move the migration code to a separate utils class.

Seemed appropriate to move them into `WPStoreUtils` in 065612806d59b9475060360b9c5733b52bea7720, don't think there's a better utils class.

> We should only do the migration if there isn't a token in the AccountStore.

Done in 7ee99210a34fae9950d84bcbdce5dcddb3ab713b

**Comment by maxme** [comment]
> Seemed appropriate to move them into WPStoreUtils in 0656128, don't think there's a better utils class.

IMO, we should move that in WPLegacyMigrationUtils - so all this mess and hard coded table and field names are contained in one place. We'll also have to write some code for migrating self hosted without Jetpack.

**Comment by maxme** [comment]
Re https://github.com/wordpress-mobile/WordPress-Android/pull/4430#issuecomment-237253688

We must dispatch these 3 events when the migration step happens, cause you get an empty app just after migration (no site, nothing in the Me tab, etc...)

**Comment by tonyr59h** [comment]
> Migration test: checkout an old version (5.6.1 for instance), sign in a wpcom account - then checkout this branch, build and install, make sure you're still logged in.

Checked the upgrade from `5.6.1` and it works without hiccups. Signing out then restarting the app showed that token wasn't being deleted after migrating [here](https://github.com/wordpress-mobile/WordPress-Android/pull/4430/files#diff-8d36829d091466f13662bc49aa3bc756R66) (works now).

I'm going to verify the upgrade paths from older tokens as well.

**Comment by maxme** [comment]
Looks great, but if the network request takes some time to update the site list, the user could get an empty "My site" screen, which is pretty scary. I think we don't have other solution than adding some sort of a progress dialog that would stop when we receive the onSiteChanged and onAccountChanged events.

The other solution would be: start the Sign In process like a new user signing in but with the auth token prefilled (ie, the user don't have to enter his username or password). Sorry I just noticed this issue.

**Comment by tonyr59h** [comment]
> I think we don't have other solution than adding some sort of a progress dialog that would stop when we receive the onSiteChanged and onAccountChanged events.

Seems like the easiest solution to me.

**Comment by maxme** [comment]
I still get the bug when the sites are not synced (and I don't see the progress indicator) just after migrating from an older (non FluxC) version. What I see after the migration:

screenshot-2016-09-09_09 42 48 876

**Comment by maxme** [comment]
After closing the app and restarting it, the sites show up, but the app was stuck on the progress dialog:

screenshot-2016-09-09_09 48 10 525


I really think the SignInFragment step is the safest solution, it's well tested and we're sure we won't see weird issues like this.

**Comment by tonyr59h** [comment]
> I really think the SignInFragment step is the safest solution, it's well tested and we're sure we won't see weird issues like this.

Thanks for the suggestion, that's definitely the best spot to handle the progress dialog.

I've tested the upgrade path when signed in, not signed in, and with/without an internet connection with no issues. [Here](https://cldup.com/frJpqNs0_X.png) is an updated screenshot with the progress dialog.

**Comment by maxme** [comment]
Good job @tonyr59h! And sorry for the first suggestion that wasn't the best idea ;) - Looks better now, but can we reuse the `mProgressBarSignIn` element instead of adding a progress dialog? actually I'm pretty sure we could reuse `startProgress(String message)` and `endProgress()` methods.

**Comment by tonyr59h** [comment]
Jitccup worked itself out after a few build attempts. This one is ready for merge, @maxme.

**Comment by maxme** [comment]
Success:
- [x] Tested with a good network connection (I got a smart lock for password popup, it was a bit weird, but everything worked)
- [x] Tested without a network connection (excepted failure, and when trying again with network enabled, everything worked)

Failure:
- [x] Tested with a throttled network connection: it didn't work and was stuck on the "Updating..." progress message. Reason was an underlying network error. It's not well managed during the sign in process.

**Comment by maxme** [comment]
> Tested with a throttled network connection: it didn't work and was stuck on the "Updating..." progress message. Reason was an underlying network error. It's not well managed during the sign in process.

I think this branch is mergeable, I'll open a new issue for working on better management of errors during sign in.

**Comment by maxme** [comment]
Opened #4641 
:shipit:

**Pull request merged**